### DISCUSSION TITLE
[3.5.0-RC3] One field is instantiated multiple times

### DISCUSSION THREAD

**Discussion by sebastian-lenz** [craftcms/cms#8876] [Ideas]
### Description

Up until now, a field (meaning an instance of `craft\base\Field`) has only been created exactly one time per configured field in the control panel. In the current release candidate one configured field can be instantiated multiple times, meaning the field class constructor will be invoked multiple times with the same field record.

### Details

In previous versions this happened in `craft\services\Fields::getAllFields` in [Fields.php#L572](https://github.com/craftcms/cms/blob/develop/src/services/Fields.php#L572). All other field utility methods basically just filtered this list of fields once loaded (e.g. `craft\services\Fields::getFieldByUid` in [Fields.php#L631](https://github.com/craftcms/cms/blob/develop/src/services/Fields.php#L631))

In Craft 3.5-RC3 there is another call to the field constructor version in `craft\services\Fields::getFieldsByLayoutId` in [Fields.php#L572](https://github.com/craftcms/cms/blob/develop/src/services/Fields.php#L1113).

I am not sure whether this is an intended behavior introduced cause users now can e.g. edit the label of fields. As a field might be pretty complex (e.g. matrix fields) it seems rather unnecessary to recreate them entirely just to change the field label. Also fields might rely on the fact that they are only created once (mine did), this might lead to inconsistencies as the field instance that loads a field from the database might not be the same as the one that displays the field.

### Steps to reproduce

1. Assuming we have a fresh and clean Craft installation, create a new field in the control panel. Attach it to a new entry type.
2. Place a debugger breakpoint within the constructor function of that field.
3. Visit the entry edit page in the control panel. The debugger will halt two times as the field is created twice. One time within `craft\services\Fields::getAllFields` and one time within `craft\services\Fields::getFieldsByLayoutId`.

### Additional info

- Craft version: 3.5.0-RC3
- PHP version: 7.4.7
- Database driver & version: MySQL 5.7.29

**Comment by brandonkelly** [comment]
That actually has been the case since 3.0.0.

https://github.com/craftcms/cms/blob/0ec7014ffd69109642a4cb72a7a7f4bf5d6b5f04/src/services/Fields.php#L1027-L1049

**Comment by sebastian-lenz** [comment]
Hi @brandonkelly, thanks for having a look! You are totally right, my conclusions are wrong and the double field creation existed like this in prior versions of Craft. So let us scrap my explanation. Yet the underlying observation of the changed behavior in Craft 3.5 is still true.


### Observation

The field instance that is used to retrieves a value from the database is not the same instance that will render the input elements in the control panel. Or, in other words, `craft\base\Field::normalizeValue()` and `craft\base\Field::getInputHtml()` will be called on different field class instances for a single element and field value. In prior versions of Craft, the same field instance has been used.


### (Another) Investigation

**normalizeValue** behaves consistent across Craft 3.4 and 3.5, it uses the fields stored on field layout of the element. Basically it uses a call to `craft\services\Fields::getFieldsByLayoutId()` in both versions.

**getInputHtml** behaves different between Craft 3.4 and 3.5. Prior versions of Craft use the same field layout methods as the normalize functions to retrieve the fields.  
Craft 3.5 uses the newly introduced `FieldLayoutElementInterface` to represent the mixed contents within edit forms. The `CustomField` field layout element uses the uid of a field to retrieve it in [`CustomField::setFieldUid`](https://github.com/craftcms/cms/blob/3.5/src/fieldlayoutelements/CustomField.php#L97) which will return the field instance loaded by `craft\services\Fields::getAllFields`.

### Proposal

As the field layout elements now handle the individual configurations of fields when placed inside a field layout, I see no reason as of why fields can be loaded multiple times. The additional fetched attributes are handled by the field elements nonetheless.

**Comment by brandonkelly** [comment]
I see what’s going on… In 3.0 through 3.5, all fields are fetched and memoized centrally, accessible from `craft\services\Fields` methods such as `getAllFields()`, `getFieldById()`, `getFieldByUid()`, etc.. But those memoized fields weren’t being used when fetching an element’s field value, because that ends up going through `getFieldsByLayoutId()` which (as mentioned above) will create duplicate field instances.

In Craft 3.5, when _displaying_ the field input on an element edit page, we are now using the memoized field instance rather than the duplicate one, because `craft\fieldlayoutelements\CustomField` gets its internal field reference via `getFieldByUid()`:

https://github.com/craftcms/cms/blob/1ae80255e97e7c3f1ac607fecbaa6489a81cd8d3/src/fieldlayoutelements/CustomField.php#L97-L100

But the duplicate field was still being used to retrieve the value, since that is pulled via the element.

Not a big deal, but a little awkward, and there’s no reason for `getFieldsByLayoutId()` to be reinstantiating fields anymore, after the field layout changes in 3.5. So just updated that method, as well as `getFieldsByElementType()`, to start using the memoized fields instead of reinstantiating them.

**Comment by brandonkelly** [comment]
Actually had to revert that change, because it had some unintended consequences. Will think about a better, non-breaking way to approach this, as I do agree it would be better if we could reduce the duplicate instantiated fields.
